### DISCUSSION TITLE
How to read request body from inside a Deno application?

### DISCUSSION THREAD

**Discussion by goodmanworld** [denoland/deno#12601] [Q&A] [answered]
I am trying to write a simple signup using Deno, and the following is my code:

    async register(ctx: RouterContext) {
        const { email, password } = await ctx.request.body().value;
        let user = await User.findOne({ email });
        if (user) {
            ctx.response.status = 422;
            ctx.response.body = { message: "Email is already exist" };
            return;
        }
        const hashedPassword = hashSync(password);
        user = new User({ email, password: hashedPassword });
        await user.save();
        ctx.response.status = 201;
        ctx.response.body = {
            id: user.id,
            name: user.name,
            email: user.email
        };
    }

The problem is that, what I get from the frontend side is like below:

    {
        email: "<email>",
        password: "<password>"
    }

But it seems my backend expects something like this:

    {
        value: {
            email: "<email>",
            password: "<password>"
        }
    }

And when I try to remove that `.value` part from the end of the following line of the code and change it like below:

    const { email, password } = await ctx.request.body();

It gives me error messages like:

    Property 'email' does not exist on type 'Body'.
    Property 'password' does not exist on type 'Body'.

**Comment by timofriedlberlin** [comment]
The body is the promise which can be awaited, not the value on the body. You need something like this:

= (await request.body()).value

**Reply by goodmanworld** [reply]
> The body is the promise which can be awaited, not the value on the body. You need something like this:
> 
> = (await request.body()).value

No difference. still same error.

**Reply by bartlomieju** [reply]
```
const body = await ctx.request.body();
const json = await body.json()
const { email, password } = json.value;
```

**Reply by goodmanworld** [reply]
> ```
> const body = await ctx.request.body();
> const json = await body.json()
> const { email, password } = json.value;
> ```

I get this error in the second line:
```
Property 'json' does not exist on type 'Body'.
  Property 'json' does not exist on type 'BodyBytes'.
```

**Comment by Soremwar** [comment, marked as answer]
You can access any request body by using `const result = await ctx.request.body({ type: "json" }).value;`

However the above will only work if the content type of the incoming request is application/json

**Reply by goodmanworld** [reply]
> You can access any request body by using `const result = await ctx.request.body({ type: "json" }).value;`
> 
> However the above will only work if the content type of the incoming request is application/json

Thank you. The problem was from the frontend side that I added a contentType header to requests and it solved.  But a new problem Ihave is as following:



I want to write a simple registration function and this is my code:

auth_screen.dart:

```
  Future<void> _submit() async {
    if (!_formKey.currentState!.validate()) {
      // Invalid!
      return;
    }
    _formKey.currentState!.save();
    setState(() {
      _isLoading = true;
    });
    try {
      if (_authMode == AuthMode.Login) {
        // Log user in
        await Provider.of<Auth>(context, listen: false).login(
          _authData['email'] as String,
          _authData['password'] as String,
        );
      } else {
        // Sign user up
        await Provider.of<Auth>(context, listen: false).signup(
          _authData['email'] as String,
          _authData['password'] as String,
          
        );
      }
    } on HttpException catch (error) {
      var errorMessage = 'Authentication failed';
      print("this is the auth data");
      print(_authData);
      if (error.toString().contains('EMAIL_EXISTS')) {
        errorMessage = 'This email address is already in use.';
      } else if (error.toString().contains('INVALID_EMAIL')) {
        errorMessage = 'This is not a valid email address';
      } else if (error.toString().contains('WEAK_PASSWORD')) {
        errorMessage = 'This password is too weak.';
      } else if (error.toString().contains('EMAIL_NOT_FOUND')) {
        errorMessage = 'Could not find a user with that email.';
      } else if (error.toString().contains('INVALID_PASSWORD')) {
        errorMessage = 'Invalid password.';
      }
      _showErrorDialog(errorMessage);
    } catch (error) {
      var errorMessage = 'Could not authenticate you. Please try again later.' +
          error.toString();
      _showErrorDialog(errorMessage);
    }

    setState(() {
      _isLoading = false;
    });
  }
```
auth.dart:
```

  Future<void> signup(String email, String password) async {
    return _authenticate(email, password, 'register');
  }

  Future<void> _authenticate(
      String email, String password, String urlSegment) async {

    final url = Uri.parse('http://10.0.2.2:8000/api/$urlSegment');
    // final url = Uri.http('http://localhost:8000/api/', 'urlSegment');
    try {
      final response = await http.post(
        url,
        body: json.encode(
          {
            'email': email,
            'password': password,

            //'returnSecureToken': true,
          },
        ),
      );
      final responseData = json.decode(response.body);
      if (responseData['error'] != null) {
        throw HttpException(responseData['error']['message']);
      }
      _token = responseData['idToken'];
      _userId = responseData['localId'];
      _expiryDate = DateTime.now().add(
        Duration(
          seconds: int.parse(
            responseData['expiresIn'],
          ),
        ),
      );
      _autoLogout();
      notifyListeners();
      final prefs = await SharedPreferences.getInstance();
      final userData = json.encode(
        {
          'token': _token,
          'userId': _userId,
          'expiryDate': _expiryDate!.toIso8601String(),
        },
      );
      prefs.setString('userData', userData);
    } catch (error) {
      throw error;
    }
  }
```

This is the backend(Deno) part of the project:

auth_controller.ts:

```
async register(ctx: RouterContext) {
    const { email, password } = await ctx.request.body().value;
    let user = await User.findOne({ email });
    if (user) {
        ctx.response.status = 422;
        ctx.response.body = { message: "Email is already exist" };
        return;
    }
    const hashedPassword = hashSync(password);
    user = new User({ email, password: hashedPassword });
    await user.save();
    ctx.response.status = 201;
    ctx.response.body = {
        id: user.id,
        name: user.name,
        email: user.email
    };
}
```

And this is the `user.ts` class:
```

export default class User extends BaseModel {
    public id: string = "";
    public name: string = "";
    public email: string = "";
    public password: string = "";
    constructor({ id = "", name = "", email = "", password = "" }) {
        super();
        this.id = id;
        this.name = name;
        this.email = email;
        this.password = password;
    }
    static async findOne(params: object): Promise<User | null> {
        const user = await userCollection.findOne(params);
        if (!user) {
            return null;
        }
        return new User(User.prepare(user));
    }
    async save() {
        const _id = await userCollection.insertOne(this);
        this.id = _id;
        return this;
    }
}

```
I get the following error message when I want to test the application on Android emulator:

> type 'Null' is not a subtype of type 'String'

**Reply by Soremwar** [reply]
That looks like a flutter problem,  however have you checked that the input/output from the Deno endpoint is the right one?

**Reply by goodmanworld** [reply]
> That looks like a flutter problem, however have you checked that the input/output from the Deno endpoint is the right one?

Deno app works fine with Postman at `0.0.0.0:8000/api/register` But I don't know why it doesn't work properly when I try it using Android emulator and Flutter.

**Reply by Soremwar** [reply]
Sounds like a Flutter problem then, but I can't help you since I don't know the first thing about Dart

**Reply by ghost** [reply]
by the way this does not work now. trying to find an answer for updated version in JSR.
